docs/man/onectl.1
.TH ONECTL "" "" "User Manuals"
.SH NAME
.PP
onectl - configuration tool
.SH SYNOPSIS
.PP
onectl [\f[I]options\f[]] [\f[I]input-file\f[]]...
.SH DESCRIPTION
.PP
ONECTL is a tool designed to centralize platform configuration.
It can configure certain aspects of the system at run-time, postponed
after reboot or to show active or saved system settings.
The purpose is to ease and automate system configuration and to allow
more sophisticated control over the system.
.PP
ONECTL provides a set of commands to view, set, and automate system
settings.
.PD 0
.P
.PD
ONECTL comes with a set of plug-ins responsible to configure the system.
Each plugin has a set of options in the command line allowing the
specific configuration.
For more information check the Using Command Line Interface section
below.
.PP
For a quick overview of all settings configurable in ONECTL type the
onectl as root.
The output will be something like this
.PP
\f[B]onectl\f[]
.PD 0
.P
.PD
sys.distro
.PD 0
.P
.PD
sys.kernel
.PD 0
.P
.PD
sys.time.timezone
.PD 0
.P
.PD
sys.time.ntp.service
.PD 0
.P
.PD
sys.time.ntp.servers
.PD 0
.P
.PD
net.conf.eth5.ip
.PD 0
.P
.PD
net.conf.bonds.bond1.ip
.PD 0
.P
.PD
net.conf.eth4.ip
.PD 0
.P
.PD
net.conf.eth3.ip
.PD 0
.P
.PD
net.conf.gateway
.PD 0
.P
.PD
net.bonds
.PD 0
.P
.PD
net.devices
.PD 0
.P
.PD
net.vlans
.PD 0
.P
.PD
neighbors.names
.PD 0
.P
.PD
neighbors.conf.prometheus.ip
.PD 0
.P
.PD
neighbors.conf.prometheus.ssh
.PP
ONECTL configuration abilities will be extended in future releases with
add-on plugins.
.SH SYNOPSIS
.PP
\f[B]onectl\f[] [--help] [--list] [--dump] [--load FILE] [bind
SRC-PLUGIN DSCT-PLUGIN] [unbind SRC-PLUGIN DST-PLUGIN] [-d]
.PP
\f[B]onectl\f[] [--history] [--rollback]
.PP
\f[B]onectl sys.distro\f[] [--help] [--info] [--view actual|saved|diff]
[--showall]
.PP
\f[B]onectl sys.kernel\f[] [--help] [--info] [--view actual|saved|diff]
.PP
\f[B]onectl sys.hostname\f[] [--help] [--info] [--view
actual|saved|diff] [--set]
.PP
\f[B]onectl sys.time.timezone\f[] [--help] [--info] [--view
actual|saved|diff] [--zones]
.PP
\f[B]onectl sys.time.ntp.service\f[] [--help] [--info] [--view
actual|saved|diff] [--set]
.PP
\f[B]onectl sys.time.ntp.servers\f[] [--help] [--info] [--view
actual|saved|diff] [--set SERVER SERVER..] [--add SERVER SERVER..]
[--remove SERVER SERVER..]
.PP
\f[B]onectl net.devices\f[] [--help] [--info] [--view actual|saved|diff]
[--set ETH:STATE ETH:STATE] [--all up|down] [--up ETH ETH] [--down ETH
ETH]
.PP
\f[B]onectl net.conf.ethX.ip\f[] [--help] [--info] [--view
actual|saved|diff] [--set IPADDR/MASK] [--ip IPADDR] [--mask MASK]
.PP
\f[B]onectl net.conf.gateway\f[] [--help] [--info] [--view
actual|saved|diff] [--set GATEWAY] [--remove GATEWAY]
.PP
\f[B]onectl net.bonds\f[] [--help] [--info] [--view actual|saved|diff]
[--set bondX:ethX,ethY] [--add bondX:ethX,ethY] [--remove bondX]
.PP
\f[B]onectl net.vlans\f[] [--help] [--info] [--view actual|saved|diff]
[--set ETHX.VID1 ETHX.VID2] [--add ETHX.VID] [--remove ETHX.VID]
.PP
\f[B]onectl neighbors.names\f[] [--help] [--info] [--view
actual|saved|diff] [--set HOST HOST] [--add HOST HOST] [--remove HOST
HOST]
.PP
\f[B]onectl net.conf.hostX.ip\f[] [--help] [--info] [--view
actual|saved|diff] [--set IPADDR] [--disable]
.PP
\f[B]onectl net.conf.hostX.ssh\f[] [--help] [--info] [--exchange]
[--import] [--export] [--disable]
.SH OPTIONS
.TP
.B --help
Information for onectl usage for the specified command
.RS
.RE
.TP
.B --list
[PLUG-IN] Displays list of all available plug-ins.
If PLUG-IN or part of a plugin is specified only relevant information
will be shown.If PLUG-IN is omitted all plug-ins will be shown
.RS
.RE
.TP
.B --load
FILE Loads configuration from a FILE.
For clean machine.
FILE is created in advance in the following format: PLUG-IN = VALUE on
new line each
.RS
.RE
.TP
.B -\[en]dump
[PLUG-IN] Dumps the current configuration.
If plug-in or part of it is specified only related configuration is
shown.If [PLUG-IN] is omitted then all keys are shown.
.RS
.PP
Example:
.PD 0
.P
.PD
\f[B]onectl --dump\f[]
.PD 0
.P
.PD
net.devices = eth3:up eth6:down
.PD 0
.P
.PD
net.conf.eth4.ip = 163.165.0.3/32
.PD 0
.P
.PD
net.bonds = bond4:eth7 bond3:eth1
.PD 0
.P
.PD
net.conf.eth8.ip = 222.168.2.2/32
.PD 0
.P
.PD
net.vlans = eth3.100 eth5.200
.PD 0
.P
.PD
\f[B]onectl --dump net.conf.b\f[]
.PD 0
.P
.PD
net.conf.bonds.bond0.ip = 172.18.198.232/24
.PD 0
.P
.PD
net.conf.bonds.bond1.ip = 172.31.3.232/24
.RE
.TP
.B -\[en]show
[PLUG-IN] Displays the active and saved config and the difference
between them if any.
If PLUG-IN or part of it is specified then only relevant plug-in
configuration is shown.If PLUG-IN is omitted all configuration is shown
Similar to using the "--view diff" command of a plugin.
.RS
.RE
.TP
.B -\[en]info
[PLUG-IN] Displays information for plug-ins usage and how to perform
configuration.
If PLUG-IN or part of it is specified then only relevant information is
shown.If PLUG-IN is omitted information for all plug-ins is shown
.RS
.RE
.TP
.B --nolive, -n
Do not proceed to a live configuration of the system.
The configuration is saved and executed later after reload
.RS
.RE
.TP
.B -\[en]bind
SRC-PLUG-IN DST-PLUG-IN Links two plug-ins from the same type.
If the value of DST-PLUG-IN is changed SRC-PLUG-IN is changed also.
.RS
.RE
.TP
.B -\[en]unbind
SRC-PLUG-IN DST-PLUG-IN Removes the lin between two plug-ins from the
same type.
The oposite of --bind.
.RS
.RE
.TP
.B -\[en]history
[CHANGEID] Shows configuration change history - date,change.
If CHANGEID is specified the the onectl configuration state is dumped.
Available IDs can be viewed with onectl --history command.
History is lost on uninstall
.RS
.RE
.TP
.B -\[en]rollback
[CHANGEID] Revert configuration to a previous state identified by
CHANGEID.
If CHANGEID is not specified the config is reverted to the previous
config.
Available CHANGEIDs can be viewed with onectl --history command.
.RS
.RE
.TP
.B -d
Activate debug mode.
Place -d option in each command to be debugged
.RS
Example:
.PD 0
.P
.PD
onectl -d net.bonds --add bond0:eth4,eth5
.PD 0
.P
.PD
.RE
.TP
.B PLUG-IN
Plug-in name
.RS
.RE
.SH Advanced usage:
.TP
.B --load-plugin
Use this command to load (or reload) dynamic plugins.
.RS
.RE
.SH PLUG-IN
.TP
.B \ sys.distro
Provide information about the installed OS (Version and Profile)
.RS
.PP
--\f[B]help\f[] Command information
.PD 0
.P
.PD
--\f[B]info\f[] Provides information for the specified plug-in
.PD 0
.P
.PD
--\f[B]view actual|saved|diff\f[] Dumps plug-in configuration
.IP
.nf
\f[C]
actual\ \[en]\ \ \ \ \ Dumps\ actual\ system\ configuration\ \ 
saved\ \ \[en]\ \ \ \ \ Configured\ but\ not\ actual\ yet\ \ 
diff\ \ \ \[en]\ \ \ \ \ Difference\ between\ saved\ and\ actual\ configuration\ \ 
\f[]
.fi
.PP
--\f[B]showall\f[] Show all Operating System information-both installed
OS \[en] version
.PD 0
.P
.PD
and profile and kernel information
.RE
.TP
.B \ sys.kernel
Provide information about the running kernel
.RS
.PP
--\f[B]help\f[] Command information
.PD 0
.P
.PD
--\f[B]info\f[] Provides information for the specified plug-in
.PD 0
.P
.PD
--\f[B]view actual|saved|diff\f[] Dumps plug-in configuration
.IP
.nf
\f[C]
actual\ \[en]\ \ \ \ \ Dumps\ active\ system\ configuration\ \ 
saved\ \ \[en]\ \ \ \ \ Configured\ but\ not\ active\ yet\ \ 
diff\ \ \ \[en]\ \ \ \ \ Difference\ between\ saved\ and\ active\ configuration\ \ 
\f[]
.fi
.RE
.TP
.B \ sys.hostname
Configure system name
.RS
.PP
--\f[B]help\f[] Command information
.PD 0
.P
.PD
--\f[B]info\f[] Provides information for the specified plug-in
.PD 0
.P
.PD
--\f[B]view actual|saved|diff\f[] Dumps plug-in configuration
.IP
.nf
\f[C]
actual\ \[en]\ \ \ \ \ Dumps\ active\ system\ configuration\ \ 
saved\ \ \[en]\ \ \ \ \ Configured\ but\ not\ active\ yet\ \ 
diff\ \ \ \[en]\ \ \ \ \ Difference\ between\ saved\ and\ active\ configuration\ \ 
\f[]
.fi
.PP
--\f[B]set\f[] HOSTNAME Change the name of the system
.RE
.TP
.B \ sys.time.timezone
Provides ability to configure and view the system\[aq]s timezone
.RS
.PP
--\f[B]help\f[] Command information
.PD 0
.P
.PD
--\f[B]info\f[] Provides information for the specified plug-in
.PD 0
.P
.PD
--\f[B]view actual|saved|diff\f[] Dumps plug-in configuration
.IP
.nf
\f[C]
actual\ \[en]\ \ \ \ \ Dumps\ active\ system\ configuration\ \ 
saved\ \ \[en]\ \ \ \ \ Configured\ but\ not\ active\ yet\ \ 
diff\ \ \ \[en]\ \ \ \ \ Difference\ between\ saved\ and\ active\ configuration\ \ 
\f[]
.fi
.PP
--\f[B]zones all|ZONE\f[]
.IP
.nf
\f[C]
\ Lists\ all\ available\ time\ zone\ if\ all\ command\ is\ used\ or\ a\ specific\ ZONE··
\ Used\ with\ sys.time.timezone
\f[]
.fi
.RE
.TP
.B \ sys.time.ntp.service
Provides ability to start/stop NTP service and view current status
.RS
.PP
--\f[B]help\f[] Command information
.PD 0
.P
.PD
--\f[B]info\f[] Provides information for the specified plug-in
.PD 0
.P
.PD
--\f[B]view actual|saved|diff\f[] Dumps plug-in configuration
.IP
.nf
\f[C]
actual\ \[en]\ \ \ \ \ Dumps\ active\ system\ configuration\ \ 
saved\ \ \[en]\ \ \ \ \ Configured\ but\ not\ active\ yet\ \ 
diff\ \ \ \[en]\ \ \ \ \ Difference\ between\ saved\ and\ active\ configuration\ \ 
\f[]
.fi
.PP
--\f[B]set on|off\f[]
.IP
.nf
\f[C]
Start\ or\ stop\ NTP\ service.
Used\ with\ sys.time.ntp.service\ \ 
\f[]
.fi
.PP
--\f[B]status\f[]
.IP
.nf
\f[C]
Show\ service\ status:\ if\ enabled\ or\ disable/if\ synchronized
or\ not.\ Used\ with\ sys.time.ntp.service
\f[]
.fi
.RE
.TP
.B \ sys.time.ntp.servers
Provides ability to configure NTP servers
.RS
.PP
--\f[B]help\f[] Command information
.PD 0
.P
.PD
--\f[B]info\f[] Provides information for the specified plug-in
.PD 0
.P
.PD
--\f[B]view actual|saved|diff\f[] Dumps plug-in configuration
.IP
.nf
\f[C]
actual\ \[en]\ \ \ \ \ Dumps\ active\ system\ configuration\ \ 
saved\ \ \[en]\ \ \ \ \ Configured\ but\ not\ active\ yet\ \ 
diff\ \ \ \[en]\ \ \ \ \ Difference\ between\ saved\ and\ active\ configuration\ \ 
\f[]
.fi
.PP
--\f[B]set\f[] SERVER SERVER..
.IP
.nf
\f[C]
Deletes\ all\ servers\ set\ in\ /etc/ntp.conf\ and\ adds
the\ list\ specified\ in\ the\ command\ in\ the\ same\ order\ \ 
\f[]
.fi
.PP
--\f[B]add\f[] SERVER SERVER..
.IP
.nf
\f[C]
Adds\ the\ list\ of\ servers\ specified\ to\ the\ already\ saved\ ones··
\f[]
.fi
.PP
--\f[B]remove\f[] SERVER SERVER..
.IP
.nf
\f[C]
Removes\ servers\ from\ configuration\ \ 
\f[]
.fi
.RE
.TP
.B \ net.devices
Activate or deactivate physical network interfaces
.RS
.PP
--\f[B]help\f[] Command information
.PD 0
.P
.PD
--\f[B]info\f[] Provides information for the specified plug-in
.PD 0
.P
.PD
--\f[B]view actual|saved|diff\f[] Dumps plug-in configuration
.IP
.nf
\f[C]
actual\ \[en]\ \ \ \ \ Dumps\ active\ system\ configuration\ \ 
saved\ \ \[en]\ \ \ \ \ Configured\ but\ not\ active\ yet\ \ 
diff\ \ \ \[en]\ \ \ \ \ Difference\ between\ saved\ and\ active\ configuration\ \ 
\f[]
.fi
.PP
--\f[B]set\f[] ETH:STATE ETH:STATE
.IP
.nf
\f[C]
Activate\ or\ deactivate\ physical\ network\ interfaces\ in\ this\ format:
ETH:STATE\ ETH:STATE.\ \ 
Example:\ --set\ eth0:up\ eth1:down\ \ 
\f[]
.fi
.PP
--\f[B]all up|down\f[]
.IP
.nf
\f[C]
Activate\ or\ deactivate\ all\ interfaces\ at\ once
\f[]
.fi
.PP
--\f[B]up ETH ETH\f[]
.IP
.nf
\f[C]
Activate\ one\ or\ more\ interfaces\ \ 
\f[]
.fi
.PP
--\f[B]down ETH ETH\f[]
.IP
.nf
\f[C]
Activate\ one\ or\ more\ interfaces\ \ 
\f[]
.fi
.RE
.TP
.B \ net.conf.ethX.ip net.conf.bonds.bondX.ip net.conf.vlans.ethX.Y.ip
Configures IP and/or mask to a physical network interfaces An interface
must first be activated before being able to proceed with its
configuration.
It is done by net.devices by activating or deactivating the interface
.RS
.PP
--\f[B]help\f[] Command information
.PD 0
.P
.PD
--\f[B]info\f[] Provides information for the specified plug-in
.PD 0
.P
.PD
--\f[B]view actual|saved|diff\f[] Dumps plug-in configuration
.IP
.nf
\f[C]
actual\ \[en]\ \ \ \ \ Dumps\ active\ system\ configuration\ \ 
saved\ \ \[en]\ \ \ \ \ Configured\ but\ not\ active\ yet\ \ 
diff\ \ \ \[en]\ \ \ \ \ Difference\ between\ saved\ and\ active\ configuration\ \ 
\f[]
.fi
.PP
--\f[B]set\f[] IPADDR/MASK | dhcp | none
.IP
.nf
\f[C]
Configure\ an\ IP\ address\ and\ or\ mask\ to\ an\ interface\ \ 
It\ is\ done\ in\ the\ following\ format:\ set\ IP/MASK\ \ 
eg:\ --set\ 192.168.1.1/24\ \ 
The\ \[aq]dhcp\[aq]\ keyword\ can\ also\ be\ used\ for\ dynamic\ IP\ configuration.\ \ 
Example:\ --set\ dhcp\ \ 
To\ unset\ an\ interface\ IP\ you\ can\ use\ either\ "0.0.0.0/0"\ or\ "none".\ \ 
\f[]
.fi
.PP
--\f[B]ip\f[] IPADDR
.IP
.nf
\f[C]
Configures\ IP\ address··
\f[]
.fi
.PP
--\f[B]mask\f[] MASK..
.IP
.nf
\f[C]
Configures\ mask\ \ 
\f[]
.fi
.RE
.TP
.B \ net.aliases
Creates alias network interfaces
.RS
.PP
--\f[B]help\f[] Command information
.PD 0
.P
.PD
--\f[B]info\f[] Provides information for the specified plug-in
.PD 0
.P
.PD
--\f[B]view actual|saved|diff\f[] Dumps plug-in configuration
.IP
.nf
\f[C]
actual\ \[en]\ \ \ \ \ Dumps\ active\ system\ configuration\ \ 
saved\ \ \[en]\ \ \ \ \ Configured\ but\ not\ active\ yet\ \ 
diff\ \ \ \[en]\ \ \ \ \ Difference\ between\ saved\ and\ active\ configuration\ \ 
\f[]
.fi
.PP
--\f[B]set\f[] ETHX:NUM ETHX:NUM..
.IP
.nf
\f[C]
Configures\ subinterfaces\ in\ this\ format\ ETH:NUM\ ETH:NUM.\ Removes\ the\ existing\ aliases
Example:\ --set\ eth0:1\ eth1:0
\f[]
.fi
.PP
--\f[B]add\f[] ETHX:NUM ETHX:NUM..
.IP
.nf
\f[C]
Add\ \ alias\ interfaces\ to\ already.\ \ 
Example:\ --add\ eth0:1\ eth1:0\ \ 
\f[]
.fi
.PP
--\f[B]remove\f[] ETHX:NUM ETHX:NUM..
.IP
.nf
\f[C]
Removes\ alias\ interfaces\ \ 
\f[]
.fi
.RE
.TP
.B \ net.conf.aliases.ethX:NUM.ip net.conf.aliases.bondX:1.ip
Configures IP and/or mask to a alias interfaces An interface must first
be activated before being able to proceed with its configuration.
It is done by net.devices by activating or deactivating the interface
.RS
.PP
--\f[B]help\f[] Command information
.PD 0
.P
.PD
--\f[B]info\f[] Provides information for the specified plug-in
.PD 0
.P
.PD
--\f[B]view actual|saved|diff\f[] Dumps plug-in configuration
.IP
.nf
\f[C]
actual\ \[en]\ \ \ \ \ Dumps\ active\ system\ configuration\ \ 
saved\ \ \[en]\ \ \ \ \ Configured\ but\ not\ active\ yet\ \ 
diff\ \ \ \[en]\ \ \ \ \ Difference\ between\ saved\ and\ active\ configuration\ \ 
\f[]
.fi
.PP
--\f[B]set\f[] IPADDR/MASK | dhcp | none
.IP
.nf
\f[C]
Configure\ an\ IP\ address\ and\ or\ mask\ to\ an\ interface\ \ 
It\ is\ done\ in\ the\ following\ format:\ set\ IP/MASK\ \ 
eg:\ --set\ 192.168.1.1/24\ \ 
The\ \[aq]dhcp\[aq]\ keyword\ can\ also\ be\ used\ for\ dynamic\ IP\ configuration.\ \ 
Example:\ --set\ dhcp\ \ 
To\ unset\ an\ interface\ IP\ you\ can\ use\ either\ "0.0.0.0/0"\ or\ "none".\ \ 
\f[]
.fi
.PP
--\f[B]ip\f[] IPADDR
.IP
.nf
\f[C]
Configures\ IP\ address··
\f[]
.fi
.PP
--\f[B]mask\f[] MASK..
.IP
.nf
\f[C]
Configures\ mask\ \ 
\f[]
.fi
.RE
.TP
.B \ net.conf.gateway

onectl/bash/onectl.bash
## ONECTL Bash completion
# For alias plugin : is used in the name escape this special character and return it back at the end

USER=$(whoami)

export COMP_WORDBREAKS=${COMP_WORDBREAKS/\:/}
GLOBAL_OPTS="--help --list --dump --load --nolive --show --info --bind --unbind --history --rollback"
SPTH='/usr/share/onectl/plugins/'
check_plugin() {
	rel_path=$(echo $plugin | sed -e "s/\./\//g")
	plug_path="$SPTH$rel_path.py"
	noactions=""
	if [ -e "$plug_path" ]; then
		# Check if it is a symlink to a template
		if [ -L "$plug_path" ]; then
			destination="$(readlink -f $plug_path)"
			xml_path="/usr/share/onectl/plugins/$rel_path.xml"
			# if there is an xml, then take the commands from the xml
			if [ -L "$xml_path" ]; then
				# get the xml type
				file_type=$(sed -n '/<\/file_type>/ {
									s/^.*<file_type>//
									s/<\/file_type>.*$//
									p
									}' $xml_path)
				if [ "ini" == "$file_type" ]; then
					# take the plugin info
					plugin_info=$(sed -n '/'"<name>$plugin"'/,/<\/plugin>/ {
							s/^.*<plugin>//
							s/<\/plugin>.*$//
							p
							}' $xml_path)
					tag="type"
					# get the type value. If a list add add and remove commands
					type_value=$(echo $plugin_info | sed -n '/.*<'"$tag"'>/,/<\/'"$tag"'>/ {
							s/^.*<'"$tag"'>//
							s/<\/'"$tag"'>.*$//
							p
							}' )
					if [ "LIST" == "$type_value" -o "INTEGER-LIST" == "$type_value" ]; then
						choice="--help --info --view  --set --add --remove --disable"
					else
						choice="--help --info --view  --set --disable"
					fi
				else
					actions=$(grep "### OPTION:" $plug_path | sed -e "s/ *$//" | grep -v "info$\|list$\|show$" | sed -e "s/.*OPTION: */--/" | sed -e "s/ *//")
					noactions=$(grep "### NO OPTION:" $plug_path | sed -e "s/ *$//" | grep -v "info$\|list$\|show$" | sed -e "s/.*OPTION: */--/" | sed -e "s/ *//")
					choice="--help --info --view  --set $actions"
				fi
			
			else
				actions=$(get_actions)
				noactions=$(grep "### NO OPTION:" $plug_path | sed -e "s/ *$//" | grep -v "info$\|list$\|show$" | sed -e "s/.*OPTION: */--/" | sed -e "s/ *//")
				choice="--help --info --view  --set $actions"
			fi
		else
			actions=$(get_actions)
			noactions=$(grep "### NO OPTION:" $plug_path | sed -e "s/ *$//" | grep -v "info$\|list$\|show$" | sed -e "s/.*OPTION: */--/" | sed -e "s/ *//")
			choice="--help --info --view  --set $actions"
		fi
	else
		choice=""
	fi
	for toremove in $noactions; do
		tmp=$choice
		choice=$(echo $tmp | sed -e "s/$toremove *//")
	done
}

get_actions() {
	actions=$(grep "### OPTION:" $plug_path | sed -e "s/ *$//" | grep -v "info$\|list$\|show$" | sed -e "s/.*OPTION: */--/" | sed -e "s/ .*//" | sed -e "s/\[[[:alnum:]]*\]//g")
	echo "$actions"
}

get_extended_actions() {
	CHECK_OPT=$1
	plg_opt_path=""
	multi_choice=""
	ACTS=$(grep "### OPTION:" $plug_path | sed -e "s/ *$//" | grep -v "info$\|list$\|show$" | sed -e "s/.*OPTION: */--/" | sed -e "s/ \[/_[/")
	for action in "$ACTS"; do
		if [ $(echo "$action" | sed -e "s/_.*//" | grep -e "^$CHECK_OPT$") ]; then
			TEST_PATH=$(echo "$action" | grep "_\[PATH\]")
			TEST_MULTI=$(echo "$action" | grep "_\[.*\]")
			if [ "$TEST_PATH" ]; then
				shorted=$(echo $action | sed -e "s/_.*//")
				plg_opt_path="$plg_opt_path $shorted"
			elif [ "$TEST_MULTI" ]; then
				multi_choice=$(echo $action | sed -e "s/.*_\[//" | sed -e "s/\].*//")
			fi
		fi
	done
}

check_option_old() {
	first_dash=$(echo $OPT_WORD | grep "^-")
	if [ "$first_dash" ]; then
		choice=$GLOBAL_OPTS
	elif [ "$OPT_WORD" == "--load" ]; then
		get_path
	else
		choice=$(find "$SPTH" | grep "\.py$" | grep -v "~$" | sed -e "s/.*plugins.//" | sed -e "s/.py$//" | sed -e "s/\//./g")
	fi
}

check_option() {
	first_dash=$(echo $OPT_WORD | grep "^-")
	if [ "$first_dash" ]; then
		choice=$GLOBAL_OPTS
	elif [ "$OPT_WORD" == "--load" ]; then
		get_path
	else
		get_plugin_list
	fi
}
get_plugin_list_old() {
	TMPWORD=$(echo $OPT_WORD | sed -e "s|\.|/|g" |sed -e "s|/$||")
	DIR_LST=$(find "$SPTH" | grep "\.py$" | grep "plugins\/$TMPWORD" | grep -v "~$" | sed -e "s/.*plugins.//" | sed -e "s/.py$//")
	if [ "$TMPWORD" ]; then
		choice=$(echo -e "$DIR_LST" | sed -e "s|\($TMPWORD/\([A-Za-z0-9]*-*_*[A-Za-z0-9]*:*[0-9]*\)*\)/.*|\1/|g" | sed -e "s|/|.|g")
	else
		choice=$(echo -e "$DIR_LST" | sed -e "s|\(\([A-Za-z0-9]*-*_*[A-Za-z0-9]*:*[0-9]*\)*\)/.*|\1/|g" | sed -e "s|/|.|g")
	fi
}

get_plugin_list() {
	TMPWORD=$(echo $OPT_WORD | sed -e "s|\.|/|g" |sed -e "s|/$||" | sed -e "s| $||")
	FULL_PATH="$SPTH$TMPWORD"
	while [ ! -d $FULL_PATH ]; do
		NEW_PATH=${FULL_PATH%\/*}
		FULL_PATH=$NEW_PATH
	done
	
	TEST_WORD="$SPTH$TMPWORD"
	KEY_LIST=$(find $FULL_PATH/ | grep "$TEST_WORD" | grep "\.py$" | grep -v "~$" | sed -e "s/.py$//")
	CONTENT_LIST=""
	TMP_PLG_PATH=${FULL_PATH##*"/usr/share/onectl/plugins"}
	PLG_PATH=$(echo "$TMP_PLG_PATH" | sed -e "s|^/||")
	for KEY in $KEY_LIST; do
		TMP_KEY=${KEY##*$FULL_PATH/}
		NEW_KEY=${TMP_KEY%%\/*}
		if [ ! "$NEW_KEY" ]; then
			NEW_KEY="$TMP_KEY"
		elif [ -d "$FULL_PATH/$NEW_KEY" ]; then
			NEW_KEY="$NEW_KEY/"
		fi
		if [ "$PLG_PATH" ]; then
			TEST=$(echo -e "$CONTENT_LIST" | grep " $PLG_PATH/$NEW_KEY ")
			if [ ! "$TEST" ]; then
				CONTENT_LIST="$CONTENT_LIST $PLG_PATH/$NEW_KEY "
			fi
		else
			# We are at root plugins
			TEST=$(echo -e "$CONTENT_LIST" | grep " $NEW_KEY ")
			if [ ! "$TEST" ]; then
				CONTENT_LIST="$CONTENT_LIST $NEW_KEY "
			fi
		fi
	done
	choice=$(echo -e "$CONTENT_LIST" | sed -e "s|/|.|g" | sed -e "s/^ //" | sed -e "s/ $//")
	if [ $(echo "$choice" | wc -w) -eq 1 ]; then
		if [ $(echo "$choice" | grep "\.$") ]; then
			# We extend the list if there is only one subdir available
			OPT_WORD=$choice
			get_plugin_list
		fi
	fi
		
	echo "#$choice#" >> /tmp/onectl-$USER.tmp
}

get_path()
{
	if [ ! "$OPT_WORD" ]; then
		choice=$(ls)
	else
		FULL_PATH=$OPT_WORD
		LOCAL_FILE=$(echo $FULL_PATH | grep "^[A-Za-z]")
		if [ "$LOCAL_FILE" ]; then
			TMP_PATH="$PWD/$FULL_PATH"
			FULL_PATH=$TMP_PATH
			TMP_WORD="$PWD/$OPT_WORD"
			OPT_WORD=$TMP_WORD
		fi
		while [ ! -d $FULL_PATH ]; do
			NEW_PATH=${FULL_PATH%\/*}
			FULL_PATH=$NEW_PATH
		done
		if [ ! "$FULL_PATH" ]; then
			FULL_PATH="/"
		fi
		CONTENT_LIST=$(find $FULL_PATH -maxdepth 1 | grep "^$OPT_WORD")
		if [ "$LOCAL_FILE" ]; then
			choice=$(echo -e "$CONTENT_LIST" | sed -e "s|^$PWD/||" | sed -e "s/ $//" | sed -e "s/^ //")
		else
			choice=$(echo -e "$CONTENT_LIST" | sed -e "s/^ //" | sed -e "s/ $//")
		fi
		if [ $(echo "$choice" | wc -w) -eq 1 ]; then
			if [ -d "$choice" ]; then
				# We extend the list if there is only one subdir available
				OPT_WORD="$choice/"
				get_path
			fi
		fi
	fi
}
_onectl() {
	local choice current plg_opt_path
	MULTI_OPTS="-n -d --nolive --show --info"
	SHOW_OPTS="--show --info --dump --list"
	case "$COMP_CWORD" in
		# If we are at the first parameter, then list all available plugins
		1)
			OPT_WORD=${COMP_WORDS[1]}
			check_option
			choice="$choice"" [--OPT]"
			;;
		
		# On the second level, list possible actions
		2)
			first_opt=$(echo ${COMP_WORDS[1]} | grep "^-")
			if [[ $MULTI_OPTS =~ (^| )${COMP_WORDS[1]}($| ) ]]; then
				OPT_WORD=${COMP_WORDS[2]}
				check_option
			elif [ -z "$first_opt" ]; then
				plugin=${COMP_WORDS[1]}
				check_plugin
			elif [ "${COMP_WORDS[1]}" == "--dump" ]; then
				OPT_WORD=${COMP_WORDS[2]}
				check_option
			elif [ "${COMP_WORDS[1]}" == "--help" ]; then
				choice=''
			elif [ "${COMP_WORDS[1]}" == "--list" ]; then
				OPT_WORD=${COMP_WORDS[2]}
				get_plugin_list
			elif [ "${COMP_WORDS[1]}" == "--bind" -o  "${COMP_WORDS[1]}" == "--unbind" ]; then
				OPT_WORD=${COMP_WORDS[2]}
				get_plugin_list
			elif [ "${COMP_WORDS[1]}" == "--load" ]; then
				OPT_WORD=${COMP_WORDS[2]}
				get_path
			elif [ "${COMP_WORDS[1]}" == "--rollback" ]; then
				choice='ID'
			elif [ "${COMP_WORDS[1]}" == "--history" ]; then
				choice='ID'
			else
				choice=""
			fi
			;;
		3)
			last_opt=$(echo ${COMP_WORDS[2]} | grep "^-")
			if [ "${COMP_WORDS[1]}" == "--load" ]; then
				OPT_WORD=${COMP_WORDS[3]}
				get_path
			elif [ "${COMP_WORDS[1]}" == "--bind" -o "${COMP_WORDS[1]}" == "--unbind" ]; then
				OPT_WORD=${COMP_WORDS[3]}
				get_plugin_list
			elif [[ $SHOW_OPTS =~ (^| )${COMP_WORDS[1]}($| ) ]]; then
				choice=''
			elif [ -z "$last_opt" ]; then
				plugin=${COMP_WORDS[2]}
				check_plugin
			elif [ "${COMP_WORDS[2]}" == "--view" ]; then
				choice='actual saved diff'
			else
				get_extended_actions ${COMP_WORDS[2]}
				if [ "$plg_opt_path" ]; then
					OPT_WORD=${COMP_WORDS[3]}
					get_path
				elif [ "$multi_choice" ]; then
					choice="$multi_choice"
				else
					choice=''
				fi
			fi
			;;
		4)
			last_opt=$(echo ${COMP_WORDS[3]} | grep "^-")
			if [ "${COMP_WORDS[1]}" == "--load" ]; then
				OPT_WORD=${COMP_WORDS[4]}
				get_path
			elif [ "${COMP_WORDS[2]}" == "--load" ]; then
				OPT_WORD=${COMP_WORDS[4]}
				get_path
			elif [ "${COMP_WORDS[3]}" == "--view" ]; then
				choice='actual saved diff'
			else
				EXTENDED=$(get_extended_actions)
				if [ "$( echo $EXTENDED | grep -e "${COMP_WORDS[3]}\( \|$\)" )" ]; then
					OPT_WORD=${COMP_WORDS[4]}
					get_path
				else
					choice=''
				fi
			fi
			;;
		5)
			last_opt=$(echo ${COMP_WORDS[4]} | grep "^-")
			if [ "${COMP_WORDS[1]}" == "--load" ]; then
				OPT_WORD=${COMP_WORDS[5]}
				get_path
			elif [ "${COMP_WORDS[2]}" == "--load" ]; then
				OPT_WORD=${COMP_WORDS[5]}
				get_path
			elif [ -z "$last_opt" ]; then
				plugin=${COMP_WORDS[4]}
			elif [ "${COMP_WORDS[4]}" == "--view" ]; then
				choice='actual saved diff'
			else
				EXTENDED=$(get_extended_actions)
				if [ "$( echo $EXTENDED | grep -e "${COMP_WORDS[4]}\( \|$\)" )" ]; then
					OPT_WORD=${COMP_WORDS[5]}
					get_path
				else
					choice=''
				fi
			fi
			;;
		*)
			words=()
			;;
	esac
	# creat the final list of choices
	current=${COMP_WORDS[COMP_CWORD]}
	COMPREPLY=( $( compgen -W '$choice' -- $current) )
}

complete -F _onectl onectl


